doc/design.md

```diff
@@ -0,0 +1,922 @@
+- Coordinate and name matching enables mathematical operations on individual table columns.
+
+##### DataObjects::MDHistoWorkspace
+
+A dataset with multi-dimensional variables `Data::Value` and `Data::Variance` in combination with a corresponding number of one--dimensional coordinate variables is a close equivalent to the current `DataObjects::MDHistoWorkspace`.
+A typical example would be three-dimensional `Data::Value` and `Data::Variance` alongside one-dimensional `Coord:Qx`, `Coord:Qy`, and `Coord::Qz`.
+
+##### DataObjects::Workspace2D
+
+A dataset with two-dimensional variables `Data::Value` and `Data::Variance` in combination with a one- or two-dimensional variable `Coord::Tof` is a close equivalent to the current `DataObjects::Workspace2D`.
+- Typically `Data::Value` and `Data::Variance` would have dimensions `Dim::Tof` and `Dim::Spectrum`.
+- If `Coord::Tof` has only `Dim::Tof`, it corresponds to a `DataObjects::Workspace2D` with shared X axis, if it also has `Dim::Spectrum` X is not shared.
+- To represent bin edges, the extent of `Coord::Tof` in `Dim::Tof` is larger by one than `Dim::Tof` of the other variables.
+- `Coord::SpectrumNumber` provides a spectrum number (currently in `API::ISpectrum`).
+- Mapping from spectra to detectors (currently in `API::ISpectrum`) can be added using `Coord::DetectorGrouping`.
+  For representing the instrument see the corresponding section below.
+- Data from `API::ExperimentInfo` such as `API::Run` can be held by zero-dimensional variables such as `Data::ExperimentLog`.
+- Masking of complete spectra or individual bins support by one- or two-dimensional variables such as `Coord::Mask`.
+
+Most basic operations have an obvious mapping to what our current basic algorithms are doing.
+
+##### Geometry::DetectorInfo, Geometry::ComponentInfo, and Geometry::Instrument
+
+`Geometry::DetectorInfo` and `Geometry::ComponentInfo` are the result of the refactoring effort as part of **Instrument-2.0**.
+Thanks to that refactoring, the underlying data is already stored in a structure of vectors, which can be mapped to variables in a dataset in an almost 1:1 manner.
+- In place of `Geometry::DetectorInfo`, several variables with `Dim::Detector` would be added to a dataset.
+  - `Coord::DetectorPosition` to store positions, `Coord::DetectorRotation` to store rotations, ...
+- The option to have multi-dimensional variables will provide support for instruments with (synchronously) scanning detectors in a natural way.
+  - Variables such as `Coord::DetectorPosition` can have the additional `Dim::DetectorScan` (in addition to `Dim::Detector`).
+  - Time intervals for the scan can be stored in an additional variable `Coord::TimeInterval`.
+- Data for `Geometry::ComponentInfo` can be mapped to variables in the dataset in a similar manner.
+- `Geometry::Instrument` still exists as part of the pre-**Instrument-2.0** legacy.
+  For the time being it can be supported by simply holding it in a zero-dimensional variable.
```

Comment:
If you have ideas about how how we expect the instrument to be propagated through datasets and algorithms without breaking sharing, I think that would be interesting to comment on. Mainly, will the propagation happen automatically at the dataset level, or is it the algorithm's responsibility?

Reply:
It happens automatically, just like the sharing works now in, e.g., `DetectorInfo`. The only remaining bit is to ensure that it can be shared between dataset with different origin, i.e., files loaded separately. Basically what `InstrumentDataService` does.